### PR TITLE
IR-115: Add support to mark a task as 'Done' and 'Undone'

### PR DESCRIPTION
**Pull request by ishugupta92**

### CODE REVIEW

File: apps/retrospective/services/task.go [hunk 1]
```diff
@@ -66,6 +66,50 @@ func (service TaskService) Get(id string, retroID string, sprintID string) (task
...
 }
 
+// MarkDone ...
+func (service TaskService) MarkDone(id string, retroID string, sprintID string) (task *retroSerializers.Task, status int, err error) {
+	db := service.DB
+	var sprint retroModels.Sprint
+	err = db.Model(&retroModels.Sprint{}).
+		Where("id = ?", sprintID).
+		Scan(&sprint).Error
+
```
Comment (anmolghosh) [inline apps/retrospective/services/task.go:76]: what if sprint doesn't exist, shouldn't 404 error be ideal then?

File: apps/retrospective/services/task.go [hunk 2]
```diff
@@ -66,6 +66,50 @@ func (service TaskService) Get(id string, retroID string, sprintID string) (task
...
+		return nil, http.StatusInternalServerError, errors.New("failed to mark the task as done")
+	}
+
+	err = db.Model(&retroModels.Task{}).
+		Where("tasks.id = ?", id).
+		Where("done_at is NULL").
+		Update("done_at", *sprint.EndDate).
+		Error
+
+	if err != nil {
```
Comment (anmolghosh) [inline apps/retrospective/services/task.go:91]: same as above

File: apps/retrospective/services/task.go [hunk 3]
```diff
@@ -66,6 +66,50 @@ func (service TaskService) Get(id string, retroID string, sprintID string) (task
 	return &tasks[0], http.StatusOK, nil
 }
 
+// MarkDone ...
+func (service TaskService) MarkDone(id string, retroID string, sprintID string) (task *retroSerializers.Task, status int, err error) {
```
Comment (anmolghosh) [inline apps/retrospective/services/task.go]: let's use taskId as param for instance, we are using `sprintID` in sprint service

File: controllers/v1/task.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ type TaskController struct {
 func (ctrl TaskController) Routes(r *gin.RouterGroup) {
 	r.GET("/", ctrl.List)
 	r.GET("/:taskID/", ctrl.Get)
+	r.POST("/:taskID/mark-done", ctrl.MarkDone)
```
Comment (anmolghosh) [inline controllers/v1/task.go]: keep single url, post for done, delete for un-done.
Also, should we call it resolve or done, chose which looks more apt to you

File: db/migrations/00015_add_role_to_member_task.go [hunk 5]
```diff
@@ -26,6 +26,8 @@ func Up00015(tx *sql.Tx) error {
 
 	gormdb.AutoMigrate(&sprintMemberTask{})
 
+	gormdb.Model(&models.Task{}).AddIndex("idx_tasks_done_at", "done_at")
```
Comment (anmolghosh) [inline db/migrations/00015_add_role_to_member_task.go]: why was this added here, should not it be in the new migration added below

**Review by anmolghosh — CHANGES_REQUESTED** [review]

File: apps/retrospective/services/task.go (hunk 3 above)
Comment (ishugupta92) [reply apps/retrospective/services/task.go]: Done

File: db/migrations/00015_add_role_to_member_task.go (hunk 5 above)
Comment (ishugupta92) [reply db/migrations/00015_add_role_to_member_task.go]: I accidentally added this here, nice catch!. Fixed

File: apps/retrospective/services/task.go (hunk 1 above)
Comment (ishugupta92) [reply apps/retrospective/services/task.go:76]: Done

File: apps/retrospective/services/task.go (hunk 2 above)
Comment (ishugupta92) [reply apps/retrospective/services/task.go:91]: Done

File: controllers/v1/task.go (hunk 4 above)
Comment (ishugupta92) [reply controllers/v1/task.go]: Done

**Review by anmolghosh — APPROVED** [review]
